### PR TITLE
Generate error for unused variables

### LINKED ISSUE #357: Make unused local variables be an error
The spec currently says nothing about unused variables, but my expectation is that this will become an error.

We should implement this now to avoid accumulating test cases with unused variables.

To make this convenient, we should first implement #356 (so you can so `_ = x;` to "use" `x`).

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose
Fixes #357

### CODE REVIEW

File: compiler/modules/front/codeGen.bal [hunk 1]
```diff
@@ -359,11 +362,42 @@ function codeGenStmts(CodeGenContext cx, bir:BasicBlock bb, Environment initialE
             env.assignments.push(...effect.assignments);
         }
     }                
+    check unusedLocalVariables(cx, env);
     int[] assignments = [];
     addAssignments(assignments, env.assignments, startRegister);
     return { block: curBlock, assignments };
 }
 
+function unusedLocalVariables(CodeGenContext cx, Environment env) returns CodeGenError? {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: There's a simpler way to do this. Pass in a parameter `Binding? limit` which is initialEnv.bindings in codeGenStmts. Then check the used field in each binding in env up to limit that is not a narrowed binding. You don't need localVarNames.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/codeGen.bal (hunk 1 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Changed

File: compiler/modules/front/codeGen.bal [hunk 2]
```diff
@@ -359,11 +360,54 @@ function codeGenStmts(CodeGenContext cx, bir:BasicBlock bb, Environment initialE
...
     }                
+    check unusedLocalVariables(cx, env, initialEnv.bindings);
     int[] assignments = [];
     addAssignments(assignments, env.assignments, startRegister);
     return { block: curBlock, assignments };
 }
 
+function unusedLocalVariables(CodeGenContext cx, Environment env, Binding? bindingLimit) returns CodeGenError? {
+    Binding? tem = env.bindings;
+    while true {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: This looks too complicated. You can stop when `tem === bindingLimit`. If we haven't reached that, then tem must be non-nil.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/codeGen.bal (hunk 2 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Fixed

File: compiler/modules/front/codeGen.bal [hunk 3]
```diff
@@ -359,11 +360,40 @@ function codeGenStmts(CodeGenContext cx, bir:BasicBlock bb, Environment initialE
...
     }                
+    check unusedLocalVariables(cx, env, initialEnv.bindings);
     int[] assignments = [];
     addAssignments(assignments, env.assignments, startRegister);
     return { block: curBlock, assignments };
 }
 
+function unusedLocalVariables(CodeGenContext cx, Environment env, Binding? bindingLimit) returns CodeGenError? {
+    Binding? binding = env.bindings;
+    while true {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: while binding !== bindingLimit

File: compiler/modules/front/codeGen.bal [hunk 4]
```diff
@@ -359,11 +360,40 @@ function codeGenStmts(CodeGenContext cx, bir:BasicBlock bb, Environment initialE
...
+        if binding === bindingLimit {
+            break;
+        }
+        else {
+            // binding is non-nil
+            Binding tem = <Binding>binding;
+            if isUnused(tem) {
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: I think you can just do `if tem.unnarrowed == () && !tem.used`. The unnarrowed binding must have been added before bindingLimit; you can just ignore the narrowed one.

**Review by jclark — CHANGES_REQUESTED** [review]

**Comment by jclark** [comment]
In order to make sure we identify all unused variables in existing tests, I suggest
- make unusedLocalVariables print out a message for each unused variable (instead of returning an error)
- run it on all tests cases other than 10-unused to identify all unused variables in tests
- fix these
- then make it return an error
- maybe make the test runner report an error if it gets an unused variable error and neither the directory nor file contain "unused"

File: compiler/testSuite/01-function/call13-e.bal [hunk 5]
```diff
@@ -1,6 +1,6 @@
 // @productions return-stmt local-var-decl-stmt int-literal
 public function main() {
-    int x = 1;
+    _ = 1;
```
Comment (jclark) [inline compiler/testSuite/01-function/call13-e.bal]: This is not an OK change!!!!! Fix like this:

```
int x = 1;
_ = x;
```

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/testSuite/01-function/call13-e.bal (hunk 5 above)
Comment (KavinduZoysa) [reply compiler/testSuite/01-function/call13-e.bal]: @jclark, since we are testing `unreachable code` error in this test case, can you please explain the reason for the above particular change? 
What I thought was, since we are testing `unreachable code`, it does not matter the statements we have before the return-stmt.

Comment (jclark) [reply compiler/testSuite/01-function/call13-e.bal]: You are changing what the test is testing.  One of the things that the original test is checking is that `1` can be assigned to a variable of type `int`. With your change, that is no longer being tested. You should only ever be adding lines that are assignments to `_`.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front/codeGen.bal (hunk 3 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Fixed

File: compiler/modules/front/codeGen.bal (hunk 4 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Fixed

File: compiler/testSuite/01-function/call13-e.bal (hunk 5 above)
Comment (KavinduZoysa) [reply compiler/testSuite/01-function/call13-e.bal]: Thank you, I understood.
Fixed

Comment (manuranga) [reply compiler/testSuite/01-function/call13-e.bal]: Five exclamation marks, the sure sign of an insane mind. ― Sir Terry Pratchett

**Review by jclark — CHANGES_REQUESTED** [review]
I did a few cases that use an `ignore` function for the same purpose as `_`. Can you also fix them to use `_`?

```
compiler/testSuite/03-list/01-e.bal:public function ignore(any x) {
compiler/testSuite/04-list/1-e.bal:function ignore(any x) { }
compiler/testSuite/04-map/01-e.bal:function ignore(any x) { }
compiler/testSuite/04-map/03-e.bal:function ignore(any x) { }
compiler/testSuite/07-array/int1-e.bal:function ignore(any x) {}
compiler/testSuite/07-array/int4-e.bal:function ignore(any x) {}
compiler/testSuite/07-array/int7-e.bal:function ignore(any x) {}
compiler/testSuite/07-map/float2-v.bal:function ignore(any x) {}
compiler/testSuite/07-map/int1-e.bal:function ignore(any x) {}
compiler/testSuite/07-map/int4-e.bal:function ignore(any x) {}
compiler/testSuite/07-map/int5-p.bal:function ignore(any x) {}
compiler/testSuite/07-map/int7-e.bal:function ignore(any x) {}
```

There are some others with a parameter type other than `any`. Let's leave them alone for now.

File: compiler/modules/front/codeGen.bal [hunk 6]
```diff
@@ -359,11 +360,25 @@ function codeGenStmts(CodeGenContext cx, bir:BasicBlock bb, Environment initialE
...
+    Binding? binding = env.bindings;
+    while binding !== bindingLimit {
+        // binding is non-nil
+        Binding tem = <Binding>binding;
+        if tem.unnarrowed == () && !tem.used {
+            return cx.semanticErr(`unused local variable ${tem.name}`);
+        }
+        binding = tem.prev;
+    }
+    return ();
```
Comment (jclark) [inline compiler/modules/front/codeGen.bal]: Don't need `return ();` here.

File: compiler/testSuite/04-misc/evalorder-v.bal [hunk 7]
```diff
@@ -14,55 +14,55 @@ function two()  returns int{
 
 public function main() {
    int mul = one() * two();        // @output 1
-                                   // @output 2
+   _ = mul;                        // @output 2
```
Comment (jclark) [inline compiler/testSuite/04-misc/evalorder-v.bal]: This is confusing because the line is doing the output. Please add your line after both comments.

**Review by jclark — CHANGES_REQUESTED** [review]
Also need to add some tests that show that variable shadowing from type narrowing is interacting correctly with this.

File: compiler/testSuite/04-misc/evalorder-v.bal (hunk 7 above)
Comment (KavinduZoysa) [reply compiler/testSuite/04-misc/evalorder-v.bal]: Fixed

File: compiler/modules/front/codeGen.bal (hunk 6 above)
Comment (KavinduZoysa) [reply compiler/modules/front/codeGen.bal]: Fixed

**Comment by KavinduZoysa** [comment]
> Also need to add some tests that show that variable shadowing from type narrowing is interacting correctly with this.

Added

**Comment by KavinduZoysa** [comment]
> I did a few cases that use an `ignore` function for the same purpose as `_`. Can you also fix them to use `_`?
> 
> ```
> compiler/testSuite/03-list/01-e.bal:public function ignore(any x) {
> compiler/testSuite/04-list/1-e.bal:function ignore(any x) { }
> compiler/testSuite/04-map/01-e.bal:function ignore(any x) { }
> compiler/testSuite/04-map/03-e.bal:function ignore(any x) { }
> compiler/testSuite/07-array/int1-e.bal:function ignore(any x) {}
> compiler/testSuite/07-array/int4-e.bal:function ignore(any x) {}
> compiler/testSuite/07-array/int7-e.bal:function ignore(any x) {}
> compiler/testSuite/07-map/float2-v.bal:function ignore(any x) {}
> compiler/testSuite/07-map/int1-e.bal:function ignore(any x) {}
> compiler/testSuite/07-map/int4-e.bal:function ignore(any x) {}
> compiler/testSuite/07-map/int5-p.bal:function ignore(any x) {}
> compiler/testSuite/07-map/int7-e.bal:function ignore(any x) {}
> ```
> 
> There are some others with a parameter type other than `any`. Let's leave them alone for now.

Changed

File: compiler/testSuite/01-function/assign12-v.bal [hunk 8]
```diff
@@ -0,0 +1,12 @@
+// @productions local-var-decl-stmt if-else-stmt string-literal
```
Comment (jclark) [inline compiler/testSuite/01-function/assign12-v.bal]: Why is this added?

File: compiler/testSuite/04-map/01-e.bal [hunk 9]
```diff
@@ -1,8 +1,6 @@
-// @productions map-type-descriptor mapping-constructor-expr any function-call-expr local-var-decl-stmt int-literal
+// @productions map-type-descriptor mapping-constructor-expr any local-var-decl-stmt int-literal
 public function main() {
     map<any> m = {};
     int i = 2;
-    ignore(m[i]); // @error
+    _ = m[i]; // @error
```
Comment (jclark) [inline compiler/testSuite/04-map/01-e.bal]: Leave this one as is.

**Review by jclark — COMMENTED** [review]
We need some `-v` tests. For example: unused param is ok; unused function is ok; using only narrowed variable is ok. Using v only in `v[i] = e` is OK.

File: compiler/testSuite/01-function/assign12-v.bal (hunk 8 above)
Comment (KavinduZoysa) [reply compiler/testSuite/01-function/assign12-v.bal]: This is a `-v` test to check narrowed used variable. I added it under assignment tests. Now it is moved to `10-unused` test folder.

File: compiler/testSuite/04-map/01-e.bal (hunk 9 above)
Comment (KavinduZoysa) [reply compiler/testSuite/04-map/01-e.bal]: Changed

**Comment by KavinduZoysa** [comment]
> We need some `-v` tests. For example: unused param is ok; unused function is ok; using only narrowed variable is ok. Using v only in `v[i] = e` is OK.

Added
